module/zfs/vdev_disk.c

```diff
@@ -83,16 +84,31 @@ vdev_bdev_mode(int smode)
 #endif /* HAVE_OPEN_BDEV_EXCLUSIVE */
 
 static uint64_t
-bdev_capacity(struct block_device *bdev)
+bdev_capacity(struct block_device *bdev, boolean_t wholedisk)
 {
 	struct hd_struct *part = bdev->bd_part;
-
-	/* The partition capacity referenced by the block device */
-	if (part)
-		return (part->nr_sects << 9);
-
-	/* Otherwise assume the full device capacity */
-	return (get_capacity(bdev->bd_disk) << 9);
+	if (part) {
+		if (wholedisk) {
+			/* Total number of sectors on the device */
+			uint64_t sectors = get_capacity(bdev->bd_disk);
+			/* Sectors used by the EFI partition and partition alignment */
+			uint64_t used = EFI_MIN_RESV_SIZE +
+			    NEW_START_BLOCK +
+			    PARTITION_END_ALIGNMENT;
+			if (sectors <= used)
+				return (0);
+			/* Sectors available to the vdev */
+			uint64_t available = sectors - used;
+			/* Bytes available to the vdev */
+			return (available << 9);
```

Comment:
Even though `wholedisk` is set we should add an additional sanity check to verify we never expand in to an existing partition.  You can use `bdev->bd_contains` to access the full block device (/dev/sda) instead of the partition.

The on-disk partition table also somehow needs to be updated to reflect the new expanded size.  Otherwise, there's no visibility from user space that this space is in use and it will be reported by `gparted` and other tools as free space.  Since the kernel does support online resizing of partitions have you considered instead updating the on-disk partition table and then triggering a rescan with `__blkdev_reread_part()`.  That should result in an updated `part->nr_sects` which reflects the new expanded size.

Reply:
1. Sure - I'll look into that.

2. It's my understanding that the partition table should not be updated until zfs has decided to actually use the new space (i.e. the user runs `zpool online -e`). At which point the partition is relabeled by `zpool_relabel_disk` in `zpool_vdev_online`. I verified that `parted` showed unused, extra space after the new space was detected but showed the new sized partition (with no unused space) after I ran `zpool online -e.` I tried this work flow on illumos for comparison and the behavior seems similar: `fdisk` did not see the new space until it was explicitly onlined. Granted, this was all tested with `autoexpand` set to off, but I can test `autoexpand` cases as well. 

I did consider triggering a scan from `vdev_disk_open` with `__blkdev_reread_part()` but I don't think that makes sense given what I wrote above. Also, initial attempts at that failed because the `BLKRRPART` ioctl returns `EBUSY` for open devices making it unusable for hot-expansion.

Reply:
> 2. `zpool online -e`

Ahh I see.  OK that reasonable, I'd forgotten the `zpool online -e` would do the relabel,  In that case I think the only remaining issue is to the make sure we don't accidentally stomp on an existing partition (1. above).

Reply:
Cool. Should that check be something like `bdev == bdev->bd_contains' - to make sure that this is the `whole` disk?  I'm a little unsure if that makes sense, because a zfs `wholedisk` configured device should actually have two partitions. Do we want to check that there are only two partitions on the vdev, one of which is the EFI partition size? Or is there some other check I'm not thinking of?

Reply:
I think we'd ideally like to scan the kernel's version of the partition table for conflicting partitions.  But it looks like the functions needed to easily do this aren't available.  So why don't we go with your suggestion and simply verify `bdev != bdev->bd_contain` and `bdev->bd_contain->bd_part_count == 2`.  The primary EFI partition and legacy partition 9.  

Speaking of which, is partition 9 recreated when `zpool online -e` is run and moved to the new end of the device?  What is the illumos behavior here?

Reply:
Sounds good. 

Yeah - in both linux and illumos `efi_use_whole_disk` computes the difference in device size and shifts partition 9 back that much.

Reply:
OK, thanks for verify that.  Also please make sure you address the two style warnings when you refresh this.  `make checkstyle` should pass without warning when run locally.

```
./module/zfs/vdev_disk.c: 94: line > 80 characters
./module/zfs/vdev_disk.c: 105: line > 80 characters
```

Reply:
Will do. 

After looking around some more I'm not sure if `bd_part_count` is what we want for this - in `linux/fs.h` it says that `bd_part_count` is "number of times partitions within this device have been opened." Do we know that partitions will only be referenced once each in zfs? I haven't found anything else to get the number of partitions per bdev though.

Reply:
My mistake.  It looks like we want the `gendisk->part_tbl` but we're sadly not going to be able to use any of the helpers.  So you're going to need to take the bdev `bd_mutex` when checking the partition count.

> Do we know that partitions will only be referenced once each in zfs?

Actually we do.  The primary ZFS partition will be opened exclusively which prevents any other process on the system from opening it.  Although that doesn't help is in this case where we're worried about overwriting a partition we don't own.

Reply:
So `ASSERT(bdev->bd_contain->bd_part_count == 2)` isn't entirely useless - if it fails we know that either this `bdev` has no partitions or that zfs or someone else has referenced more than the expected two. However, it would miss the case when we've mistakenly passed `wholedisk` and the second partition actually belongs to someone else?

Reply:
Not exactly, we can only expect the first partition to be open and ZFS will never open the second one.  Checking `gendisk->part_tbl` would be the best way to check for extra partitions, but that's still awkward.

Since no additional space will actually be used until `zpool online -e` is run why don't we move this additional sanity check there instead.  Checking this from user space in `zpool_relabel_disk()` avoids this entire problem.  Then remove `ASSERT(bdev->bd_contain->bd_part_count == 2)`.

Reply:
Alright, it makes sense to have sanity checks where we're actually going to take the space.
My idea would be to read the `efi_label` from the file descriptor for the disk in `zpool_relabel_disk` and check that `efi_nparts == 2` Does that make sense to you?

Reply:
I think we can do a little better now that we have access to the full partition table.  Instead of checking that `efi_nparts == 2`, I was thinking you could iterate over the `efi_parts[]` array and verify that no partition starts in the expansion region (aside from the one we expect).

This has the advantage of not breaking auto-expansion for the case where a partition has been added before the primary ZFS data partition.  This might sound like an unlikely case but there are systems which use this space for a grub or boot partition.  And I could see this becoming a more common configuration in the future.

Reply:
One other thing.  With your fix in place we should be able to re-enable the `zpool_expand_001_pos`  and `zpool_expand_003_neg.ksh` tests case which are currently being skipped.

The expansion tests currently create pools on zvols, which can occasionally result in a deadlock, so you might need to adapt them to use a scsi_debug or loopback device.  There are some helper function for this in `blkdev.shlib`.  I believe you'll also need to create a zedlet script for the zed to run `zpool online -e` when the expansion event is received.

Reply:
I'm not sure of the benefit of this kind of check. We only called `zpool_relabel_disk` if `wholedisk` is true. My understanding is that if `wholedisk` is true, ZFS believes that it is the only user of the device - let me know if this is wrong. 

`zpool_relabel_disk` is the only caller of `efi_use_whole_disk` and in `efi_use_whole_disk` we assume that the reserved partition is the very last partition and that the zfs data partition is the one that comes just before that. That means that if we reach `zpool_relabel_disk` with a configuration like the following, it would be misinterpreted with `B` as `reserved` and `reserved` as `ZFS data`. The only way to tell something was wrong would be to notice that there are more than 2 partitions - i.e. we were wrong about the value of `wholedisk`. 
```
-------------------------------------------------------------------------------
|      A     |            ZFS Data            | reserved |        B          |     
-------------------------------------------------------------------------------
```
In the following case where `efi_use_whole_disk` gets the correct values for `ZFS data` and `reserved`, a check which verified the only partition in the expanded region is `reserved` partition would by true by definition since it's already been defined as the last partition. 

```
---------------------------------------------------------------------
|      A        |            ZFS Data                 | reserved   |     
---------------------------------------------------------------------
```
My core concern is that we identify partitions based on their relative position to each other and to the end of the partition table, so checking that one only extends into the other doesn't seem meaningful. Should this type of configuration be valid for a ZFS wholedisk setup? 

Are there other issues that you think we could catch here?

===============
I'll take a look at the zpool expand tests and try and evaluate how much work it would be to get them operational - if it ends up being really involved I might not be able to get to them for awhile.

Reply:
>  if wholedisk is true ZFS believes that it is the only user of the device - let me know if this is wrong.

This is definiitely what ZFS believes, _unfortunately_ that doesn't necessarily make it true.  The end goal is just to add some logic to verify this is the case before doing the expansion so we do no harm.  Even if the admin did something like expand device and manually create a new partition+ext4 filesystem in the `B` space from your example above.

> My core concern is that we identify partitions based on their relative position to each other and to the end of the partition table.

That's definitely not ideal, but it looks like we can improve things in this area.  Since the `reserved` partition is created as part of `zpool create`, when `wholedisk` is set, we have a [little more information](https://github.com/zfsonlinux/zfs/blob/master/lib/libzfs/libzfs_pool.c#L4504) we can use to uniquely indentify it.  Then we won't have to rely on it being the last partition.  We know:

  * `vtoc->efi_parts[].p_size == EFI_MIN_RESV_SIZE`, and
  * `vtoc->efi_parts[].p_tag = V_RESERVED`, and
  * will always be partition number 9.

Reply:
Thanks for the clarification - I was starting to doubt everything... I agree that we're treating partitions a little more fast and loose than would be idea.  

Ah, ok, cool! So in `efi_use_whole_disk` we could add checks that the partition we're treating as the the `reserved` partition (the last one we find in the table) is actually the one we created. If so, then we're safe to expand because there's no next partition. If not, we should give up because the `wholedisk` assumption is no longer valid and we have no idea what we might be expanding into.